frame/uniques/src/functions.rs

```diff
@@ -277,4 +289,69 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 		Ok(())
 	}
+
+	pub fn do_set_royalties(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		new_royalties: Perbill,
+	) -> DispatchResult {
+		let collection =
+			Collection::<T, I>::get(&collection_id).ok_or(Error::<T, I>::UnknownCollection)?;
+		ensure!(collection.owner == sender, Error::<T, I>::NoPermission);
+
+		CollectionRoyaltiesOf::<T, I>::try_mutate(collection_id, |maybe_royalties| {
+			// royalties can only be decreased
+			if let Some(ref royalties) = maybe_royalties {
+				ensure!(
+					new_royalties.is_zero() || new_royalties < *royalties,
+					Error::<T, I>::RoyaltiesCantBeIncreased
+				);
+			}
+
+			let royalties = maybe_royalties.get_or_insert_with(|| Perbill::zero());
+			*royalties = new_royalties.clone();
+
+			Self::deposit_event(Event::RoyaltiesChanged {
+				collection: collection_id,
+				royalties: new_royalties,
+			});
+
+			Ok(())
+		})
+	}
+
+	fn process_royalties(
+		amount: ItemPrice<T, I>,
+		royalties: Perbill,
+		source: &T::AccountId,
+		collection_id: &T::CollectionId,
+		item_id: &T::ItemId,
+	) -> Result<ItemPrice<T, I>, DispatchError> {
+		let mut amount_left = amount.clone();
+
+		if !royalties.is_zero() {
+			let collection =
+				Collection::<T, I>::get(&collection_id).ok_or(Error::<T, I>::UnknownCollection)?;
```

Comment:
do you have to fetch the collection again here?

Instead, you can pass the already fetched collection from the outer function, and then you dont need to read storage again